### PR TITLE
 [FLINK-33317][runtime] Add cleaning mechanism for initial configs to reduce the memory usage

### PR DESCRIPTION
**Pull request by 1996fanrui**
## What is the purpose of the change

This PR is FLINK-33317, it's a subtask of FLINK-33315. See FLINK-33315 and FLINK-33317 for details.

## Brief change log

Clean up the StreamConfig#SERIALIZEDUDF after getStreamOperatorFactory.

## Verifying this change

This change added tests and can be verified as follows:

  - Added StreamConfigTest#cleanSerializedUDF

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by 1996fanrui** [comment]
Hi @pnowojski @RocMarshal , would you mind helping take a look this PR in your free time? Thanks a lot

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:809]: nit: can we mark removal of those keys with some special value instead, so that we could throw a meaningful error in case of someone trying to read an already cleared value?

**Review by pnowojski — APPROVED** [review]
one comment, apart of that lgtm :)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java
Comment (RocMarshal) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:810]: Do we want to release this part of memory faster here?

If so, Maybe we could call some native methods so that the JVM can reach the safe point faster. For example, calling the `Thread. sleep (0L)` method.
Of course, whether this portion of memory can be reclaimed as quickly as possible largely depends on the implementation of the JVM.
Just a minor comment~

**Review by RocMarshal — COMMENTED** [review]
Hi, @1996fanrui Thank you very much for the analysis and fix.
I left a comment. PTAL in your free time.

**Review by RocMarshal — APPROVED** [review]

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java
Comment (1996fanrui) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:809]: Thanks @pnowojski for the quick response!

It's definitely a good suggestion, throwing a error is easy to let developer find this improvement if any developer get them in the future.

How about adding a `private final transient Set<String> removedKeys` to save them? If so, we don't need to get bytes from config and check whether it's removedValue in advance.

Of course, these 2 solution are similar, if you prefer store a special value, I can accept it.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java
Comment (1996fanrui) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:810]: Thanks @RocMarshal for the quick response!

> Do we want to release this part of memory faster here?

If the heap memory isn't enough, the GC can be happened, and these objects can be released. So I'm not sure whether it's needed.

Comment (RocMarshal) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:810]: Thanks @1996fanrui  for the reply.

Just to reiterate this comment,

As you know, there is a hypothetical premise and a consensus premise:

1. Assuming that this part of the memory is large (sorry, I am not familiar with this part)
2. The JVM will perform GC at a secure point; GC needs to be at a safe point, but it may not necessarily trigger GC when there is a safe point. So even if the JVM reaches the safe point, it only means that there are more conditions for reaching GC, and the probability of GC triggering is higher.

Based on the above two points:

Assuming the code runs here, but the JVM has not reached the safe point, GC cannot be performed temporarily. Continuing to run will also take up some space. When GC is performed, more memory needs to be reclaimed and the latency will be longer. If we can make the JVM reach a safe point early, the probability of early GC will increase, which can minimize GC latency or allocate the latency to multiple GCs.

However, this type of case is extreme and the conditions are also unstable. Let's turn our attention to the merger of PR :)
Thank you again for your patience~

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:809]: Both are fine by me :)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java
Comment (1996fanrui) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:810]: Thanks @RocMarshal for the detailed analysis.

The reason I don't want to call the extra code is: all jobs call clearInitialConfigs(), but most jobs' UDFs aren't very large, so the less overhead is better here. (Of course the overhead is proposed by you is not huge).

And for large job, I think current strategy may be enough. If the problem you describe does occur, we can solve it in time. WDYT?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java
Comment (1996fanrui) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java]: Thanks @pnowojski for the feedback, updated.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java
Comment (RocMarshal) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java]: SGTM +1~

**Comment by 1996fanrui** [comment]
Hi @pnowojski @RocMarshal , some unit tests failed when `getStreamOperatorFactory` throwing an exception after `SERIALIZED_UDF` is removed.

I researched it for a while and found that it takes some time to refactor related tests. However, I don't have enough time this week because `flink-kubernetes-operator` will be released recently. And one autoscaler feature is nice to finished before this release, and this PR belongs to flink-1.19, it has enough time to develop. So I will work on `flink-kubernetes-operator` first, and back to this PR after that. 

Sorry for the delay,  and thanks a lot for your quick review, have a good time to you.

**Comment by 1996fanrui** [comment]
Hey @pnowojski @RocMarshal , I have fixed the test, would you mind helping double check? Thanks

**Comment by pnowojski** [comment]
Hey @1996fanrui , my 👍 still stands :)

**Comment by 1996fanrui** [comment]
> Hey @1996fanrui , my 👍 still stands :)

Thank you, I rebased master just now, let's wait for the CI. and I will merge it after CI is green.

**Pull request merged**
